data.md

```diff
@@ -545,50 +545,52 @@ The strings to connect needs to be unescaped before concatenated, because the `u
-The function interprets the range of bytes as little-endian.
+    rule #setBytesRange(_, START, _ ) => .Bytes
+      requires START <Int 0
 
-```k
-    syntax Int ::= #getRange(Bytes, Int, Int) [function, functional, smtlib(getRange)]
- // ----------------------------------------------------------------------------------
-    rule                       #getRange( _, ADDR, WIDTH) => 0                                                                       requires notBool (0 <Int WIDTH andBool 0 <=Int ADDR)
-    rule [getRange-Positive] : #getRange(BM, ADDR, WIDTH) => #get(BM, ADDR) +Int (#getRange(BM, ADDR +Int 1, WIDTH -Int 1) *Int 256) requires          0 <Int WIDTH andBool 0 <=Int ADDR
+    syntax Bytes ::= #setRange ( Bytes , Int , Int , Int ) [function, functional, smtlib(setRange)]
+ // -----------------------------------------------------------------------------------------------
+    rule                       #setRange(BM, ADDR, VAL, WIDTH) => BM
+      requires notBool (0 <Int WIDTH andBool 0 <=Int VAL andBool 0 <=Int ADDR)
+
+    rule [setRange-Positive] : #setRange(BM, ADDR, VAL, WIDTH) => #setBytesRange(BM, ADDR, Int2Bytes(WIDTH, VAL, LE))
+      requires 0 <Int WIDTH andBool 0 <=Int VAL andBool 0 <=Int ADDR
 ```
 
-`#get` looks up a key in a map, defaulting to 0 if the map does not contain the key.
-`#set` sets a key in a map, extending the map when necessary, but only when setting non-zero values.
-`#set` is total made total by ignoring invalid setting operations, i.e. trying to set an non-byte value or a negative key.
+- `#getBytesRange(BM, START, WIDTH)` reads off `WIDTH` elements from `BM` beginning at position `START` (padding with zeros as needed).
+- `#getRange(BM, START, WIDTH)` reads off `WIDTH` elements from `BM` beginning at position `START`, and converts it into an unsigned integer. The function interprets the range of bytes as little-endian.
 
 ```k
-    syntax Int   ::= #get (Bytes, Int     ) [function, functional, smtlib(bytesGet)]
-    syntax Bytes ::= #set (Bytes, Int, Int) [function, functional, smtlib(bytesSet)]
- // --------------------------------------------------------------------------------
-    rule [get-Existing] : #get(BM, KEY) => BM [KEY] requires         KEY inBytes BM
-    rule                  #get(BM, KEY) => 0        requires notBool KEY inBytes BM
+    syntax Bytes ::= #getBytesRange ( Bytes , Int , Int ) [function, functional]
+ // ----------------------------------------------------------------------------
+    rule #getBytesRange(_, START, WIDTH) => .Bytes
+      requires notBool (WIDTH >=Int 0 andBool START >=Int 0)
+
+    rule #getBytesRange(BM, START, WIDTH) => substrBytes(padRightBytes(BM, START +Int WIDTH, 0), START, START +Int WIDTH)
+      requires START >=Int 0 andBool WIDTH >=Int 0 andBool START <Int lengthBytes(BM)
+
+    rule #getBytesRange(_, _, WIDTH) => padRightBytes(.Bytes, WIDTH, 0) [owise]
```

Comment:
Why is this one needed, can't you just use the first rule for this case as well?

Reply:
Well, I guess it makes sense from a semantic viewpoint. But I'd prefer if we used an explicit condition, like
`requires 0 <=Int START andBool 0 <=Int WIDTH andBool  lengthBytes(BM) <=Int START`

Reply:
done